drivers/adc/adc_ads1x1x.c

```diff
@@ -0,0 +1,662 @@
+			break;
+		case ADC_GAIN_2:
+			config |= ADS1X1X_CONFIG(PGA, ADS1X1X_CONFIG_PGA_2048);
+			break;
+		case ADC_GAIN_4:
+			config |= ADS1X1X_CONFIG(PGA, ADS1X1X_CONFIG_PGA_1024);
+			break;
+		case ADC_GAIN_8:
+			config |= ADS1X1X_CONFIG(PGA, ADS1X1X_CONFIG_PGA_512);
+			break;
+		case ADC_GAIN_16:
+			config |= ADS1X1X_CONFIG(PGA, ADS1X1X_CONFIG_PGA_256);
+			break;
+		default:
+			return -ENOTSUP;
+		}
+	} else {
+		/* no programable gain ampilier, so only allow ADC_GAIN_1 */
+		if (channel_cfg->gain != ADC_GAIN_1) {
+			return -ENOTSUP;
+		}
+	}
+
+	config |= ADS1X1X_CONFIG(MODE,
+				 ADS1X1X_CONFIG_MODE_SINGLE_SHOT); /* Only single shot supported */
+
+	config |= ADS1X1X_CONFIG(COMP_QUE, ADS1X1X_CONFIG_COMP_QUEUE_NONE); /* disable comparator */
+
+	return ads1x1x_write_reg(dev, ADS1X1X_REG_CONFIG, config);
+}
+
+static int ads1x1x_validate_buffer_size(const struct adc_sequence *sequence)
+{
+	size_t needed = sizeof(int16_t);
+
+	if (sequence->options) {
+		needed *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int ads1x1x_validate_sequence(const struct device *dev, const struct adc_sequence *sequence)
+{
+	const struct ads1x1x_config *config = dev->config;
+
+	if (sequence->resolution != config->resolution) {
+		return -ENOTSUP;
+	}
+
+	if (sequence->channels != BIT(0)) {
+		return -ENOTSUP;
+	}
+
+	if (sequence->oversampling) {
+		return -ENOTSUP;
+	}
+
+	return ads1x1x_validate_buffer_size(sequence);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat_sampling)
+{
+	struct ads1x1x_data *data = CONTAINER_OF(ctx, struct ads1x1x_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct ads1x1x_data *data = CONTAINER_OF(ctx, struct ads1x1x_data, ctx);
+
+	data->repeat_buffer = data->buffer;
+
+	ads1x1x_start_conversion(data->dev);
+
+	k_sem_give(&data->acq_sem);
+}
+
+static int ads1x1x_adc_start_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	int rc;
+	struct ads1x1x_data *data = dev->data;
+
+	rc = ads1x1x_validate_sequence(dev, sequence);
+	if (rc != 0) {
+		return rc;
+	}
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int ads1x1x_adc_read_async(const struct device *dev, const struct adc_sequence *sequence,
+				  struct k_poll_signal *async)
+{
+	int rc;
+	struct ads1x1x_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, async ? true : false, async);
+	rc = ads1x1x_adc_start_read(dev, sequence);
+	adc_context_release(&data->ctx, rc);
+
+	return rc;
+}
+
+static int ads1x1x_adc_perform_read(const struct device *dev)
+{
+	int rc;
+	struct ads1x1x_data *data = dev->data;
+	const struct ads1x1x_config *config = dev->config;
+
+	k_sem_take(&data->acq_sem, K_FOREVER);
```

Comment:
I think it would be more clear it this was put outside of `ads1x1x_adc_perform_read()`, as this is actually a waiting to perform the result read.

Reply:
ok, it can probably be put in the function a stack above it in within `ads1x1x_acquisition_thread` while loop before the read.